types/condition.go

```diff
@@ -28,28 +31,35 @@ func getUnknownCondition(conditionType string) longhorn.Condition {
 	return condition
 }
 
-func SetConditionAndRecord(conditions map[string]longhorn.Condition, conditionType string, conditionValue longhorn.ConditionStatus,
-	reason, message string, eventRecorder record.EventRecorder, obj runtime.Object, eventtype string) map[string]longhorn.Condition {
+func SetConditionAndRecord(conditions []longhorn.Condition, conditionType string, conditionValue longhorn.ConditionStatus,
+	reason, message string, eventRecorder record.EventRecorder, obj runtime.Object, eventtype string) []longhorn.Condition {
+
+	_, condition := GetCondition(conditions, conditionType)
 
-	condition := GetCondition(conditions, conditionType)
 	if condition.Status != conditionValue {
 		eventRecorder.Event(obj, eventtype, conditionType, message)
 	}
 	return SetCondition(conditions, conditionType, conditionValue, reason, message)
 }
 
-func SetCondition(originConditions map[string]longhorn.Condition, conditionType string, conditionValue longhorn.ConditionStatus, reason, message string) map[string]longhorn.Condition {
-	conditions := map[string]longhorn.Condition{}
+func SetCondition(originConditions []longhorn.Condition, conditionType string, conditionValue longhorn.ConditionStatus, reason, message string) []longhorn.Condition {
+	conditions := []longhorn.Condition{}
 	if originConditions != nil {
 		conditions = originConditions
 	}
-	condition := GetCondition(conditions, conditionType)
+	conditionIndex, condition := GetCondition(conditions, conditionType)
 	if condition.Status != conditionValue {
 		condition.LastTransitionTime = util.Now()
 	}
 	condition.Status = conditionValue
 	condition.Reason = reason
 	condition.Message = message
-	conditions[conditionType] = condition
+
+	if conditionIndex == -1 {
+		conditions = append(conditions, condition)
+	} else {
+		conditions[conditionIndex] = condition
+	}
+
```

Comment:
Do you think we could check the `condition.Status == longhorn.ConditionStatusUnknown` rather than return the index `-1` and check the `index == -1`?

Reply:
Seems OK after quick check.